Full static typing for `torch.distributions`

randolf-scholz opened the pull request:
Fixes #144196
Extends #144197  #144106  #144110

## Open Problems /// LSP violations

- [ ] `mixture_same_family.py`: `cdf` and `log_prob` violate LSP (argument named `x` instead of `value`).
    - suggestion: Imo these kinds of methods should make use of positional-only parameters, at least in base classes.
- [ ] `exp_family.py`: LSP problem with `_log_normalizer` (parent class requires `(*natural_params: Tensor) -> Tensor`, subclasses implement `(a: Tensor, b: Tensor) -> Tensor`).
  - suggestion: change parent class signature to `(natural_params: Tuple[Tensor, ...]) -> Tensor`. While this is BC breaking, (a) this is a private method, i.e. implementation detail, and (b) [no one other than torch seems to overwrite it](https://github.com/search?q=%22def+_log_normalizer%28self%2C+*%22&type=code&p=2)
- [ ] `constraints.py`: `dependent_property`: mypy does not apply the same special casing to subclasses of `property` as it does to `property` itself, hence the need for `type: ignore[assignment]` statements.
  - affects: `relaxed_bernoulli.py`, `relaxed_categorical.py`, `logistic_normal.py`, `log_normal.py`, `kumaraswamy.py`, `half_cauchy.py`, `half_normal.py`, `inverse_gamma.py`, `gumbel.py`, `weibull.py`.
  - suggestion: consider a construction similar to `lazy_property` in `distributions/utils`.
- [ ] `constraints.py` public interface not usable as type hints.
  - More crisp design would likely have one class per constraints, instead of using a mix of classes and instances.
  - suggestion: Add 1 class per constraint in the public interface; this can be subclasses of the existing ones.
  - As a workaround, I currently added a bunch of `TypeAlias`-variants, but that is likely not the best solution.
- [ ] `transforms.py`: `_InverseTransform.with_cache` violates LSP.
  - suggestion: change `with_cache` to return `_InverseTransform`.
- [ ] `test_distributions.py`: One test uses [`Dist.arg_constraints.get`](https://github.com/pytorch/pytorch/blob/3649e2e7bde8ff06f6fe6ac4168e879e9e4f5c0a/test/distributions/test_distributions.py#L6753), hence assumes `arg_constraints` is a class-attribute, but the base class `Distribution` defines it as a `@property`.
- [ ] `test_distributions.py`: One test uses [`Dist.support.event_dim`](https://github.com/pytorch/pytorch/blob/3649e2e7bde8ff06f6fe6ac4168e879e9e4f5c0a/test/distributions/test_distributions.py#L1431), hence assumes `support` is a class-attribute, but the base class `Distribution` defines it as a `@property`.
- [ ]  `test_distributions.py`: Multiple tests use [`dist.cdf(float)`](https://github.com/pytorch/pytorch/blob/3649e2e7bde8ff06f6fe6ac4168e879e9e4f5c0a/test/distributions/test_distributions.py#L3517), but the base class annotates  `cdf(Tensor) -> Tensor`.
    - suggestion: replace float values with tensors in test, unless floats should be officially supported. Note that floats are nonsensical for [multivariate distributions](https://en.wikipedia.org/wiki/Cumulative_distribution_function#Multivariate_case), so supporting it would probably require introducing a subclass for univariate distributions.
- [ ]  `test_distributions.py`: Multiple tests use [`dist.log_prob(float)`](https://github.com/pytorch/pytorch/blob/3649e2e7bde8ff06f6fe6ac4168e879e9e4f5c0a/test/distributions/test_distributions.py#L4066), but the base class annotates  `log_prob(Tensor) -> Tensor`.

## Notes

- `__init__.py`: use `+=` instead of `extends` ([ruff PYI056](https://docs.astral.sh/ruff/rules/unsupported-method-call-on-all/))
- `binomial.py`: Allow `float` arguments in `probs` and `logits` (gets used in tests)
- `constraints.py`: made `_DependentProperty` a generic class, and `_DependentProperty.__call__` polymorphic.
- `constraint_registry.py`: Made `ConstraintRegistry.register` a polymorphic method, checking that the factory is compatible with the constraint.
- `constraint_registry.py`: Needed to add `type: ignore` comments to functions that try to register multiple different constraints at once.
  - maybe split them up?
- `dirichlet.py`: `@once_differentiable` is untyped, requires `type: ignore[misc]` comment.
- `dirichlet.py`: `ctx: Any` could be replaced with `ctx: FunctionContext`, however, the type lacks the `saved_tensors` attribute.
- `distribution.py`: `Distribution._get_checked_instance` Accessing `"__init__"` on an instance is unsound, requires `type: ignore` comment.
- `distribution.py`: Changed `support` from `Optional[Constraint]` to `Constraint` (consistent with the existing docstring, and several functions in tests rely on this assumption)
- `exp_family.py`: small update to `ExponentialFamily.entropy` to fix type error.
- `independent.py`: fixed type bug in `Independent.support`.
- `multivariate_normal.py`: Added `type: ignore` comments to `_batch_mahalanobis` caused by[^1].
- `relaxed_bernoulli.py`: Allow float temperature argument (used in tests)
- `relaxed_categorical.py`: Allow float temperature argument (used in tests)
- `transforms.py`: Needed to change `ComposeTransform.__init__` signature to accept `Sequence[Transform]` rather than just `list[Transform]` (covariance!)
- `transformed_distribution.py`: Needed to change `TransformedDistribution.__init__` signature to accept `Sequence[Transform]` rather than just `list[Transform]` (covariance!)
- `transformed_distribution.py`: `TransformedDistribution.support` is problematic, because the parent class defines it as `@property` but several subclasses define it as an attribute, violating LSP.
- `von_mises.py`: fixed `result` type being initialized as `float` instead of `Tensor`.
- `von_mises.py`: `@torch.jit.script_if_tracing` is untyped, requires `type: ignore[misc]` comment.
- `von_mises.py`: Allow float `loc` and `scale` (used in tests)

[^1]: `torch.Size` is not correctly typed, causing `mypy` to think `Size + Size` is `tuple[int, ...]` instead of `Size`, see <https://github.com/pytorch/pytorch/issues/144218>.


cc @fritzo @neerajprad @alicanb @nikitaved @ezyang @malfet @xuzhao9 @gramster

randolf-scholz committed:
torch/distributions/: type annotations
+ removed '## mypy: allow-untyped-defs' comments

randolf-scholz committed:
constraints.py: type annotations

randolf-scholz committed:
von_mises.py: fixed type errors in _eval_poly

randolf-scholz committed:
independent.py: fixed bug in Independent.support

randolf-scholz committed:
exp_family.py: fixed bug in ExponentialFamily.entropy

randolf-scholz committed:
torch/distributions/: annotated attributes

randolf-scholz commented:
@pytorchbot label "module: typing"

randolf-scholz commented:
@pytorchbot label "module: distributions"

randolf-scholz commented:
@pytorchbot label "release notes: python_frontend"

randolf-scholz committed:
utils.py: type annotations

randolf-scholz committed:
transforms.py: type annotations

randolf-scholz committed:
kl.py: type annotations

randolf-scholz committed:
__init__.py: use `+=` instead of `extends` (ruff PYI056)

randolf-scholz commented on torch/distributions/__init__.py:
```diff
@@ -169,4 +169,4 @@
     "transform_to",
 ]
-__all__.extend(transforms.__all__)
+__all__ += transforms.__all__
```
See notes in my OP, this is a ruff PYI rule.

randolf-scholz replied:
https://docs.astral.sh/ruff/rules/unsupported-method-call-on-all/

randolf-scholz committed:
test/distributions/: type annotations

randolf-scholz committed:
lintrunner fixes
+ removed unused 'type: ignore' comments

randolf-scholz reviewed (commented):
More accurate would be using some sort of polymorphic mapping like

```python
class _KL_REGISTRY_TYPE(Protocol):
    def __iter__(self) -> tuple[type[Distribution], type[Distribution]]: ...
    def __getitem__(self, key: tuple[type[P], type[Q]], / ) -> _KL[P, Q]: ...
    def __setitem__(self, key: tuple[type[P], type[Q]], value: _KL[P, Q], /) -> None: ...
    def __delitem__(self, key: tuple[type[Distribution], type[Distribution]], /) -> None: ...
    def clear(self) -> None: ...
```

but likely this overcomplicates things unnecessarily

randolf-scholz committed:
allow float arguments in '__init__' methods.

Skylion007 commented:
@randolf-scholz Still interested in merging this?

randolf-scholz commented:
@Skylion007 Yes, this was quite a bit of work, and it would be a shame if it goes to waste...

As I wrote in my last comment [and in the OP](https://github.com/pytorch/pytorch/pull/144219#issue-2769405685https://github.com/pytorch/pytorch/pull/144219#issue-2769405685), there are a few remaining open problems that mostly stem from [LSP violations](https://en.wikipedia.org/wiki/Liskov_substitution_principle). It would be good to get some feedback on these.

Also, @malfet suggested in the other PR (the one adding `__init__` signatures), which was also rather large, to split it up into multiple PRs to lessen the review burden. However, I am not sure how to do that reasonably. Is there a tool to automate this? Maybe [git-explode](https://github.com/aspiers/git-explode)?

Skylion007 commented on torch/distributions/constraints.py:
```diff
@@ -206,16 +214,19 @@ def support(self):
     def __init__(
         self,
-        fn: Optional[Callable[..., Any]] = None,
+        fn: Optional[Callable[[T], R]] = None,
```
Why not ParamSpec here? T isn't used elsewhere?

randolf-scholz replied:
`ParamSpec` does not make sense here, since this is a property-decorator, and properties are usually not supposed to take arguments. (note that `T` is actually the type the property gets bound to)

Skylion007 commented on torch/distributions/transformed_distribution.py:
```diff
@@ -48,19 +50,21 @@ class TransformedDistribution(Distribution):
     """
 
-    arg_constraints: dict[str, constraints.Constraint] = {}
+    arg_constraints: dict[str, Constraint] = {}
```
not a classvar right?

Skylion007 replied:
Do we need to type these as typing.ClassVar?

randolf-scholz replied:
`wishart.py` also has the line:

```python
self.arg_constraints["df"] = constraints.greater_than(event_shape[-1] - 1)
```

which would not be allowed if `arg_constraints` were a `ClassVar`. At the same time, this line appears in the tests:

```python
values = [
    self._perturb_tensor(
        value, Dist.arg_constraints.get(key, constraints.real)
    )
    for key, value in zip(keys, values)
]
```

which does assume that it is a `ClassVar`.

Skylion007 reviewed (approved):
I have some nits, but this is definitely an improvement for what was there before.

randolf-scholz commented on torch/distributions/kl.py:
```diff
@@ -38,17 +38,28 @@
-
 __all__ = ["register_kl", "kl_divergence"]
 
+P = TypeVar("P", bound=Distribution)
```
The codebase seems to be inconsistent with respect to that, but more often than not uses private variables. Personally, I strongly prefer non-private, because it makes hints that show up for instance with pylance more readable. Moreover, when support for 3.11 is dropped and PEP 695 is used there is really no reason anymore to use an underscore prefix.

randolf-scholz replied:
But for the purposes of this PR, I am fine with changing it if necessary.

rec replied:
We're the ones setting a lot of the standards for type naming!

I personally like the `_`.

But "The names, they convey nothing!" :-D

```
DistributionT1 = TypeVar("DistributionT1", bound=Distribution)
DistributionT2 = TypeVar("DistributionT2", bound=Distribution)
DistributionT3 = TypeVar("DistributionT3", bound=Distribution)
DistributionT4= TypeVar("DistributionT3", bound=Distribution)
```

and 

```
DistributionBinaryFunc: Callable[[DistributionT1, DistributionT2], Tensor]
```

randolf-scholz commented on torch/distributions/kl.py:
```diff
@@ -38,17 +38,28 @@
+
 
-def register_kl(type_p, type_q):
+def register_kl(type_p: type[P], type_q: type[Q]) -> _KL_Decorator:
```
That would be nice, but I do not see how to do it currently, I think it requires [HKTs](https://github.com/python/typing/issues/548). That's because below you have some cases where the same function gets decorated multiple times like

```python
@register_kl(Normal, Beta)
@register_kl(Normal, ContinuousBernoulli)
@register_kl(Normal, Exponential)
@register_kl(Normal, Gamma)
@register_kl(Normal, Pareto)
@register_kl(Normal, Uniform)
def _kl_normal_infinity(
    p: Normal, q: Union[Beta, ContinuousBernoulli, Exponential, Gamma, Pareto, Uniform]
) -> Tensor:
    return _infinite_like(p.loc)
```

So, the fist decorator must return `Callable[[Normal, Union[Beta, ContinuousBernoulli, Exponential, Gamma, Pareto, Uniform]], Tensor]`, otherwise the next decorator will cause a type error.

What would be ideal would be something like

```python
class _KL_Decorator[P, Q](Protocol):
    def __call__[P2: P, Q2: Q](self, arg: _KL[P2, Q2], /) -> _KL[P2, Q2]: ...


def register_kl(type_p: type[P], type_q: type[Q]) -> _KL_Decorator[P, Q]:
```

But this requires HKTs as the polymorphism of `_KL_Decorator` needs to be upper bounded by it's type variables `P` and `Q`

randolf-scholz replied:
Actually with `mypy==1.15` similar issues crop up in the `constraint_registry.py` file, it seems one also needs to loosen the `Factory` type hint in a similar manner.

randolf-scholz commented on torch/distributions/negative_binomial.py:
```diff
@@ -83,7 +80,7 @@ def expand(self, batch_shape, _instance=None):
         return new
 
-    def _new(self, *args, **kwargs):
+    def _new(self, *args: Any, **kwargs: Any) -> Tensor:
```
Possibly, but all this function does is call `TensorBase.new` which currently has an `*args: Any` overload. Really what would be needed here is the ability to reference the signature of another function, which is currently not a feature supported by the type system.

randolf-scholz committed:
Update torch/distributions/transforms.py

Co-authored-by: Aaron Gokaslan <aaronGokaslan@gmail.com>

randolf-scholz committed:
allow tensor input in interval-constraints (batched usage)

randolf-scholz committed:
updated constraint_registry

randolf-scholz commented on torch/distributions/uniform.py:
```diff
@@ -30,11 +31,13 @@ class Uniform(Distribution):
 
     # TODO allow (loc,scale) parameterization to allow independent constraints.
-    arg_constraints = {
+    arg_constraints: dict[str, Constraint] = {
```
Thinking about it, the way it is currently set up it probably should be. In fact, the way it is used in `wishart.py` is actually a clear bug, because instantiating a second Wishart distribution ends up modifying the constraints on the first instance.

```python
import torch
from torch.distributions import Wishart

w1 = Wishart(torch.tensor([5]), covariance_matrix=torch.eye(4))
print(w1.arg_constraints["df"])  # GreaterThan(lower_bound=3)
Wishart(torch.tensor([2]), covariance_matrix=torch.eye(2))
print(w1.arg_constraints["df"])  # GreaterThan(lower_bound=1)
```

Though to be honest I find the design a bit awkward since it requires using the `DependentProperty` on distributions where the constraints depend on the parameters.

randolf-scholz replied:
Opened #154355

randolf-scholz replied:
It should be noted that the base class actually defines `arg_constraints` as a `@property`, so it is really weird to me that for example `Uniform` defines it as a `ClassVar` with these `constraints.dependent` entries, instead of just following the base class and using a property.

randolf-scholz replied:
@Skylion007 I wrote a patch that changes `arg_constraints` to `@property` for `Wishart` and `Uniform`; for the remaining distributions we can use a `ClassVar` then.

randolf-scholz committed:
rebase

randolf-scholz committed:
added type-ignore to test_support_attributes

randolf-scholz committed:
make arg_constraints a ClassVar when parameter independent

randolf-scholz committed:
Made  a ClassVar for distributions where it is parameter independent.

randolf-scholz committed:
added some missing attribute annotations + formatting

rec commented:
@skylion: Sigh, the delta is greater than 2k lines, and this makes the "sanity check" test fail. 

```
2025-05-27T18:27:27.4803708Z Your PR is 3049 LOC which is more than the 2000 maximum
2025-05-27T18:27:27.4804568Z allowed within PyTorch infra. PLease make sure to split up
2025-05-27T18:27:27.4805366Z your PR into smaller pieces that can be reviewed.
2025-05-27T18:27:27.4806075Z If you think that this rule should not apply to your PR,
2025-05-27T18:27:27.4806753Z please contact @albanD or @seemethere.
```

It's generally pretty easy to split typing pull requests.

Starting a full review now, don't split before that! 🙂

rec reviewed (commented):
Whew, that's a big one!

I read every line though.

Thanks for doing this all!

randolf-scholz commented on torch/distributions/pareto.py:
```diff
@@ -28,7 +30,14 @@ class Pareto(TransformedDistribution):
     """
 
-    arg_constraints = {"alpha": constraints.positive, "scale": constraints.positive}
+    arg_constraints: ClassVar[dict[str, Constraint]] = {
```
Yes, that would be a nice enhancement. Thinking about it, it should probably even be `Final[ClassVar[SomeTypedDict]]`, but that would require running `mypy` with `--python-version 3.13`. (currently 3.11 in `mypy.ini`)

randolf-scholz commented on test/distributions/test_distributions.py:
```diff
@@ -132,28 +144,35 @@
-def pairwise(Dist, *params):
+class _SupportSampling(Protocol):
+    def sample(self, sample_shape: _size = torch.Size()) -> Tensor:
+        ...
```
When I submitted earlier verisons of this PR, the lintrunner was still moving these to the next line, I think it got upgraded in the meantime :smile:

randolf-scholz commented on test/distributions/test_distributions.py:
```diff
@@ -1266,7 +1287,9 @@ def _check_forward_ad(self, fn):
 
-    def _check_log_prob(self, dist, asset_fn):
+    def _check_log_prob(
+        self, dist: Distribution, asset_fn: Callable[Concatenate[int, ...], None]
```
Defining partial function signatures can be really handy, I wish this was even better supported (for instance when writing Callback Protocols)

randolf-scholz commented on test/distributions/test_distributions.py:
```diff
@@ -4062,10 +4129,10 @@ def tril_cholesky_to_tril_corr(x):
                 )
             self.assertEqual(log_probs[0], log_probs[1])
-            invalid_sample = torch.cat([sample, sample.new_ones(1, dim)], dim=0)
+            invalid_sample = torch.cat([sample, sample.new_ones(1, dim)], dim=0)  # type: ignore[possibly-undefined]
```
Yep, `mypy` doesn't understand that the loop is guaranteed to finish at least one iteration.

randolf-scholz commented on test/distributions/test_distributions.py:
```diff
@@ -5853,9 +5946,11 @@ def test_kl_lowrank_multivariate_normal_batched(self):
         for (p, _), (_, q) in self.finite_examples:
-            if type(p) == type(q) and issubclass(type(p), ExponentialFamily):
+            if type(p) == type(q) and isinstance(p, ExponentialFamily):
+                assert isinstance(p, ExponentialFamily)
```
Yes, this one is redundant. The `assert isinstance(q, ExponentialFunation)` is needed though because `mypy` is not able to make the inference that that if-clause implies this.

randolf-scholz commented on test/distributions/test_transforms.py:
```diff
@@ -38,10 +40,14 @@
+
+
+def get_transforms(cache_size: int) -> list[Transform]:
+    transforms: list[Transform] = [
```
Yes, this is kind of redundant, though in the past there have been cases where type-checkers got slow with large literal lists. I may have done this out of habit because I also use `pyright` a lot, and `pyright` can have difficulties with large literal lists.

(This is because `pyright` combines with `Union` unlike `mypy`, which tries to find a common supertype. And if there are too many union members, `pyright` tends to give up and uses `Unknown` instead)

randolf-scholz commented on torch/distributions/bernoulli.py:
```diff
@@ -34,20 +35,23 @@ class Bernoulli(ExponentialFamily):
     Args:
-        probs (Number, Tensor): the probability of sampling `1`
-        logits (Number, Tensor): the log-odds of sampling `1`
+        probs (float or Tensor): the probability of sampling `1`
```
`Number` was changed from `numbers.Number` to `torch.types.number` in https://github.com/pytorch/pytorch/issues/144788, and `torch.types.Number` is currently just `bool | int | float`.

If we want to support numpy scalars in static typing, the right thing to do would be to use the `SupportsFloat` protocol.

randolf-scholz commented on torch/distributions/constraints.py:
```diff
@@ -178,7 +182,11 @@ def is_dependent(constraint):
 
 
-class _DependentProperty(property, _Dependent):
+T = TypeVar("T", contravariant=True)
```
As I wrote in the other comment, personally I prefer without the underscore because I feel like it only adds visual noise when you see the signature popping up in the IDE tooltip (for instance with pylance in vscode).

But if you want, I can change it.

randolf-scholz commented on torch/distributions/constraints.py:
```diff
@@ -206,16 +214,19 @@ def support(self):
+        property.__init__(self, fn)
+        _Dependent.__init__(self, is_discrete=is_discrete, event_dim=event_dim)
+
+    _T2 = TypeVar("_T2", contravariant=True)
```
Right, I should be consistent with these. I think here I used private because they are defined in the class and not the module scope, but really that shouldn't matter.

randolf-scholz committed:
added suggestions from review

randolf-scholz commented on torch/distributions/__init__.py:
```diff
@@ -171,4 +171,4 @@
     "register_kl",
     "transform_to",
 ]
-__all__.extend(transforms.__all__)
```
I don't think the `*-` notation you propose is supported: https://typing.python.org/en/latest/spec/distributing.html#library-interface-public-and-private-symbols

though the type spec seems to indicate that `+= submodule.__all__` is fine, so the ruff check I was basing this change on may be outdated.

randolf-scholz committed:
Merge branch 'pytorch:main' into distributions_full_typing

randolf-scholz commented:
@rec I implemented most of your suggestions. 

So, splitting it up I think it would make most sense to first make a PR for `constraints` and `constraints_registry`, and
then `distributions` and `transforms`, since the annotations for other modules depend on these.

Regarding `constraints`, I added all these type-aliases at the bottom. The CI complained that these are not listed in `__all__`, so should I just add them to `__all__`?

Alternatively, one could just make the classes they are pointing to public, but it's not my decision to make.

randolf-scholz committed:
constraints.py: added attribute annotations

randolf-scholz committed:
constraints.py: made independent/mixturesamefamily generic

randolf-scholz committed:
constraints.py: added type-aliases to __all__

randolf-scholz committed:
Merge branch 'pytorch:main' into distributions_full_typing
